trl/data_utils.py

```diff
@@ -60,6 +60,49 @@ def is_conversational(example: Dict[str, Any]) -> bool:
     return False
 
 
+def _merge_same_role_messages(messages):
+    """
+    Merge messages that share the same role into a single message.
+
+    Args:
+        messages (List[Dict[str, str]]): List of messages, where each message is a dictionary with keys "role" and
+            "content".
+
+    Returns:
+        List[Dict[str, str]]: List of messages with the same role merged into a single message.
+
+    Example:
+
+        ```python
+        >>> messages = [
+        ...     {"role": "user", "content": "What color is the sky?"},
+        ...     {"role": "assistant", "content": "Let me think..."},
+        ...     {"role": "assistant", "content": "It is blue."}
+        ... ]
+        >>> _merge_same_role_messages(messages)
+        [{'role': 'user', 'content': 'What color is the sky?'},
+         {'role': 'assistant', 'content': 'Let me think...\nIt is blue.'}]
+        ```
+    """
+    if not messages:
+        return []
+
+    merged_messages = [messages[0]]  # Initialize with the first message
+
+    for i in range(1, len(messages)):
+        current_message = messages[i]
+        previous_message = merged_messages[-1]
+
+        if current_message["role"] == previous_message["role"]:
+            # Merge the content of the current message into the previous one
+            previous_message["content"] += "\n" + current_message["content"]
```

Comment:
Joining with `"\n"` might lead to problems when trying to split back the different steps or find the end of each step if one of the messages already contain a new line character.

Reply:
One solution for this would be to use a `step_separator` that the user woud need to set. We could verify that none of the `messages` contain this `step_separator`.

Reply:
Why would you need to "split back"?

Reply:
Sorry, it is not really splitting back the steps but finding the last token of each step (for the `Process-supervised Reward Model` training). That will cause problems if the new line character is already contained in one of the steps.

We need to find the last token of each steps as PRMs are trained "to predict the correctness of each step after the last token in each step" (https://arxiv.org/pdf/2305.20050)

Reply:
I agree with @gaetanlop - we should only concatenate the messages and leave it up to the user to insert the appropriate delimiters like `\n`

Reply:
I don't think we need to "find the last token of each step" since the steps are already separated in the dataset. During the training step, you'll get a batch with the prompt, and the completions steps. Example (pseudo code):

```python
# in the train step
import torch
from itertools import accumulate

# Batch data
batch = {
    "prompt": torch.tensor([1, 2, 3, 4]),
    "completions": [torch.tensor([5, 6, 7]), torch.tensor([8, 9, 10, 11])]
}

# Get the separator indexes
completion_lengths = torch.tensor([len(step) for step in batch["completions"]])
indexes = torch.tensor([len(batch["prompt"]) + i - 1 for i in accumulate(completion_lengths)])  # [6, 10]

# Concatenate the completions
completion = torch.cat(batch["completions"])  # [5, 6, 7, 8, 9, 10, 11]

# Concatenate prompt and completion
input = torch.cat([batch["prompt"], completion])  # [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11]

# Get the output from the reward model
output = reward_model(input)
scores = output[indexes]  # [0.345, 0.897]
```

Reply:
As a consequence of my previous comment, we don't use the `_merge_same_role_messages` function. I'm having this deleted accordingly.

Reply:
Done in b1d9f2d7b383b9c2bcf728beae2f4d686c6f3b6a